### ISSUE TITLE
Apartment doesn't work with ActionController::Live

### ISSUE THREAD

**Issue by jordan-brough** [rails-on-services/apartment#304]
When you add [ActionController::Live](https://api.rubyonrails.org/classes/ActionController/Live.html) to a controller, the controller code doesn't run in the correct tenant.

This appears to be because ActionController::Live uses a separate thread:
> The final caveat is that your actions are executed in a separate thread than the main thread.

(See https://github.com/rails/rails/blob/v7.1.5.1/actionpack/lib/action_controller/metal/live.rb#L36-L38)

In order to get `ActionController::Live` to work, I had to override [new_controller_thread](https://github.com/rails/rails/blob/v7.1.5.1/actionpack/lib/action_controller/metal/live.rb#L364-L370) in my controller in my app:

```ruby
def new_controller_thread
  # using the elevator directly because it doesn't seem to get triggered with ActionController::Live
  tenant = CustomElevator.new(nil).parse_tenant_name(request)
  super { Apartment::Tenant.switch(tenant) { yield } }
end
```

**Comment by mnovelo** [comment]
Ok, good to know. Sounds like you're able to handle this without any changes to Apartment?

**Comment by jordan-brough** [comment]
Yes but it would be great if Apartment wired that up automatically. I can take a pass at putting together a PR if that sounds ok.

**Comment by mnovelo** [comment]
This wasn't included in v3.2.0, but I hope to add such support in v4.0.0. Your help would be greatly appreciated! https://github.com/rails-on-services/apartment/discussions/312

**Comment by macfanatic** [comment]
I wonder if we can introduce support easily enough with something like the following as an included initializer in the engine, that wouldn't require any app changes and automatically modify this required method when spawning the new thread.

```ruby
# support defined in engine
module Apartment
  # to be included in an ActionController::Live controller
  module Live
    def new_controller_thread
      tenant = CustomElevator.new(nil).parse_tenant_name(request)
      super { Apartment::Tenant.switch(tenant) { yield } }
    end
  end
end

# initializer in engine
ActiveSupport.on_load(:action_controller) do
  module Live
    include ::Apartment::Live
  end
end
```

Ignore the naming, and I didn't test this out. But perhaps not a huge amount of work to support.

**Comment by AlessandroTolomio** [comment]
Hello everyone!  

I'm implementing a stream solution for an AI feature in my multi-tenant application, and luckily, I found this thread. Everything seemed to be working with the proposed solution, but occasionally, the database schema doesn’t stay set correctly.  

I’m trying to figure it out, but since it's an intermittent issue, debugging isn't easy.  

I'm using **Puma** with the following configuration:  
- **Min threads:** 5  
- **Max threads:** 5  
- **Workers:** 6  

It can run smoothly **10-15 times in a row**, and then fail **5 times in a row**.  

Has anyone experienced something similar or has any suggestions?

**Comment by mnovelo** [comment]
@AlessandroTolomio, the current version of the Apartment gem does not handle multithreading very well. We have not used Pumas multi-threading because of it and instead use a single Puma thread for each instance of our application. I don't recall how the database pooling from Rails works across Puma threads.

**Comment by julik** [comment]
Forgot to mention: this is very hard to debug without running the controller in a particular manner in tests, because in tests `ActionController::Live` does a _heisenbug_ and silently switches to serving your response inline, not in a separate thread - masking the problem.
